File: resources/android/MicrophoneFunctions.kt
package com.nativephp.microphone

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.utils.NativeActionCoordinator

/**
 * Functions related to microphone recording operations
 * Namespace: "Microphone.*"
 */
object MicrophoneFunctions {

    var microphoneRecorder: MicrophoneRecorder? = null
    private var isCallbackRegistered = false

    /**
     * Register the audio permission callback with NativeActionCoordinator
     * This is called when the microphone recording is started
     */
    private fun ensureCallbackRegistered(activity: FragmentActivity, context: Context) {
        if (!isCallbackRegistered) {
            val coord = NativeActionCoordinator.install(activity)
            coord.onAudioPermissionGranted = { id, event ->
                Log.d("MicrophoneFunctions", "🎤 Audio permission granted callback - id=$id, event=$event")

                // Initialize recorder if needed
                if (microphoneRecorder == null) {
                    microphoneRecorder = MicrophoneRecorder(context)
                }

                microphoneRecorder?.start() ?: false
            }
            isCallbackRegistered = true
            Log.d("MicrophoneFunctions", "✅ Audio permission callback registered")
        }
    }

    /**
     * Start microphone recording
     * Parameters:
     *   - id: (optional) string - Unique identifier for this recording
     *   - event: (optional) string - Custom event class to dispatch when recording completes
     * Returns:
     *   - status: string - "success" or "error"
     * Events:
     *   - Fires "NativePHP\Microphone\Events\MicrophoneRecorded" when recording is stopped
     */
    class Start(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val id = parameters["id"] as? String
            val event = parameters["event"] as? String

            Log.d("MicrophoneFunctions.Start", "🎤 Starting microphone recording with id=$id, event=$event")

            // Ensure callback is registered
            ensureCallbackRegistered(activity, activity.applicationContext)

            // Delegate to NativeActionCoordinator for permission handling
            val coord = NativeActionCoordinator.install(activity)
            coord.launchMicrophoneRecorder(id, event)

            return emptyMap()
        }
    }

    /**
     * Stop microphone recording
     * Events:
     *   - Fires "NativePHP\Microphone\Events\MicrophoneRecorded" via Livewire dispatch when stopped
     */
    class Stop(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            Log.d("MicrophoneFunctions.Stop", "⏹️ Stopping microphone recording")

            val path = microphoneRecorder?.stop()

            if (path != null) {
                // Retrieve stored id and event class from SharedPreferences
                val prefs = activity.getSharedPreferences("microphone_recording", Context.MODE_PRIVATE)
                val id = prefs.getString("pending_id", null)
                val eventClass = prefs.getString("pending_event", null)
                    ?: "NativePHP\\Microphone\\Events\\MicrophoneRecorded"

                Log.d("MicrophoneFunctions.Stop", "📤 Dispatching $eventClass with path=$path, id=$id")

                // Clean up stored values
                prefs.edit()
                    .remove("pending_id")
                    .remove("pending_event")
                    .apply()

                // Create payload JSON
                val payload = org.json.JSONObject().apply {
                    put("path", path)
                    put("mimeType", "audio/m4a")
                    if (id != null) {
                        put("id", id)
                    }
                }

                // Dispatch event directly to Livewire via JavaScript (must be on main thread)
                Handler(Looper.getMainLooper()).post {
                    NativeActionCoordinator.dispatchEvent(activity, eventClass, payload.toString())
                }
            }

            return emptyMap()
        }
    }

    /**
     * Pause microphone recording
     */
    class Pause : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            Log.d("MicrophoneFunctions.Pause", "⏸️ Pausing microphone recording")
            microphoneRecorder?.pause()
            return emptyMap()
        }
    }

    /**
     * Resume microphone recording
     */
    class Resume : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            Log.d("MicrophoneFunctions.Resume", "▶️ Resuming microphone recording")
            microphoneRecorder?.resume()
            return emptyMap()
        }
    }

    /**
     * Get current recording status
     */
    class GetStatus(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            Log.d("MicrophoneFunctions.GetStatus", "📊 Getting microphone status")

            // Initialize recorder if needed to get consistent status
            if (microphoneRecorder == null) {
                Log.d("MicrophoneFunctions.GetStatus", "🔧 Initializing MicrophoneRecorder")
                microphoneRecorder = MicrophoneRecorder(context)
            }

            val status = microphoneRecorder?.getStatus() ?: "idle"
            Log.d("MicrophoneFunctions.GetStatus", "📊 Status: $status")

            val result = mapOf("status" to status)
            Log.d("MicrophoneFunctions.GetStatus", "📤 Returning: $result")

            return result
        }
    }

    /**
     * Get path to last recording
     */
    class GetRecording(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            // Initialize recorder if needed
            if (microphoneRecorder == null) {
                microphoneRecorder = MicrophoneRecorder(context)
            }
            val path = microphoneRecorder?.getLastRecording()
            return if (path != null) {
                mapOf("path" to path)
            } else {
                mapOf("path" to "")
            }
        }
    }
}

File: resources/ios/MicrophoneFunctions.swift
import Foundation

// MARK: - Microphone Function Namespace

/// Functions related to microphone recording operations
/// Namespace: "Microphone.*"
enum MicrophoneFunctions {

    // MARK: - Microphone.Start

    /// Start microphone recording
    /// Parameters:
    ///   - id: (optional) string - Unique identifier for this recording
    ///   - event: (optional) string - Custom event class to dispatch when recording completes
    /// Returns:
    ///   - status: string - "success" or "error"
    /// Events:
    ///   - Fires "Native\Mobile\Events\Microphone\MicrophoneRecorded" when recording is stopped
    class Start: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let id = parameters["id"] as? String
            let event = parameters["event"] as? String

            print("🎤 Starting microphone recording with id=\(id ?? "nil"), event=\(event ?? "nil")")

            // Store id and event for later use when stop is called
            if let id = id {
                UserDefaults.standard.set(id, forKey: "pending_microphone_id")
            }
            if let event = event {
                UserDefaults.standard.set(event, forKey: "pending_microphone_event")
            }

            MicrophoneRecorder.shared.start()

            return [:]
        }
    }

    // MARK: - Microphone.Stop

    /// Stop microphone recording
    /// Events:
    ///   - Fires "Native\Mobile\Events\Microphone\MicrophoneRecorded" via Livewire dispatch when stopped
    class Stop: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            print("⏹️ Stopping microphone recording")

            guard let path = MicrophoneRecorder.shared.stop() else {
                return [:]
            }

            // Retrieve stored id and event class from UserDefaults
            let id = UserDefaults.standard.string(forKey: "pending_microphone_id")
            let eventClass = UserDefaults.standard.string(forKey: "pending_microphone_event")
                ?? "Native\\Mobile\\Events\\Microphone\\MicrophoneRecorded"

            print("📤 Dispatching \(eventClass) with path=\(path), id=\(id ?? "nil")")

            // Clean up stored values
            UserDefaults.standard.removeObject(forKey: "pending_microphone_id")
            UserDefaults.standard.removeObject(forKey: "pending_microphone_event")

            // Build payload
            var payload: [String: Any] = [
                "path": path,
                "mimeType": "audio/m4a"
            ]
            if let id = id {
                payload["id"] = id
            }

            // Dispatch event directly to Livewire via JavaScript (same as photo/video)
            LaravelBridge.shared.send?(eventClass, payload)

            return [:]
        }
    }

    // MARK: - Microphone.Pause

    /// Pause microphone recording
    class Pause: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            print("⏸️ Pausing microphone recording")
            MicrophoneRecorder.shared.pause()
            return [:]
        }
    }

    // MARK: - Microphone.Resume

    /// Resume microphone recording
    class Resume: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            print("▶️ Resuming microphone recording")
            MicrophoneRecorder.shared.resume()
            return [:]
        }
    }

    // MARK: - Microphone.GetStatus

    /// Get current recording status
    class GetStatus: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let status = MicrophoneRecorder.shared.getStatus()
            return ["status": status]
        }
    }

    // MARK: - Microphone.GetRecording

    /// Get path to last recording
    class GetRecording: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            if let path = MicrophoneRecorder.shared.getLastRecording() {
                return ["path": path]
            } else {
                return ["path": ""]
            }
        }
    }
}